### DISCUSSION TITLE
Integration Issue with Microsoft Teams

### DISCUSSION THREAD

**Discussion by hanyEthan** [open-webui/open-webui#7627] [Issues]
- Description
  - we are having SSO integration with AzureAD (ADD), which is working fine when we open [Open WebUI] from browser window directly.
      - if the user is not previously signed-in, [Open WebUI] will automatically redirect to AAD, for logging-in, and then redirect back to authenticate [Open WebUI] with that logged-in user.
      - if the user is already authenticated with AAD, no redirection as expected, and the flow of [login/authorize/callback] will occur, without redirection
  - but under MS Teams, [Open WebUI] will be typically hosted within an iframe, and since it's MS Teams, the user will be already authenticated
    - that means, that the routine of [login/authorize/callback] is triggered, but the callback url is returning 400
    - after looking into the problem, we found that a cookie named [oui-session] is produced by [Open WebUI] in the login call, which is not secure and lax sameSite, and is expected to be used when the callback url is being called
    - we suspect that this is the reason of the error we're saying, which is just saying [{"detail":"The email or password provided is incorrect. Please check for typos and try logging in again."}]
- possible solutions
  - we're not sure, but maybe changing that cookie into a secured sameSite none could solve the issue, similar to the normal behavior without iframes
  - or do we need to modify the settings of the AAD config to handle it better ?

**Comment by julienlebutte** [comment]
Hello @hanyEthan , it seems to be the same issue as mine : https://github.com/open-webui/open-webui/discussions/7770

**Comment by Rohan-Hazari** [comment]
Hey @hanyEthan  any update?

**Comment by dawnmarch** [comment]
same issue, any update?

**Comment by arunbugkiller** [comment]
@hanyEthan, for browser based authentication with MS SSO, what is it that you followed, can you specify the redirect UrI that you configured. I have configures <open-webui-url>/oauth/microsoft/callback 

But, it is giving error 403 forbidden once user enters username and password. 

Also, I have about 126 users currently signed-up via emai and password. Do you think running the same app via SSO would keep those users ?

**Comment by dawnmarch** [comment]
@hanyEthan 
I did a test with your idea, added a layer of nginx, and then modified the cookie return, changing SameSite=Lax to SameSite=None, which can solve the problem.
The configuration in the corresponding nginx is as follows
`proxy_cookie_flags ~ Secure HttpOnly SameSite=None;`

**Comment by arunbugkiller** [comment]
@hanyEthan I was able to resolve this issue by making sure that I include the below environment variables (you can do this by editing config.py or feed as environment variables)

NABLE_OAUTH_SIGNUP=true -e WEBUI_SESSION_COOKIE_SAME_SITE=none -e WEBUI_AUTH_COOKIE_SAME_SITE=none -e ENABLE_LOGIN_FORM=False -e OAUTH_MERGE_ACCOUNTS_BY_EMAIL=true

OAUTH_MERGE_ACCOUNTS_BY_EMAIL - setting this to true would make sure that your existing users with same email ID in AD Directory are merged. 

Thanks!

**Comment by AhmadAhsan** [comment]
hi @everyone, I think I'm late, but let me share our scenario and resolution:

We had a scenario where we're trying to embed our Open WebUI application [https://www.my-web-app-url.com] in a MS Teams Application, whereas our web application and MS Teams uses AAD for user authentication.

Our main idea was just to implement an iframe and set the src attribute to our Open WebUI application. But the problem is as soon as it redirects to login.microsoft.com using iframe it is blocked for security reasons, which means iframe was not an option for us.

Moving forward we find out that we don't need iframe at all as we can set our Open WebUI URL in Teams App manifest.json file staticTabs / contentUrl property:
https://learn.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema#statictabs

With this approach our application started working fine in MS Teams App (consider it as like MS Teams works as a browser for our Open WebUI application - AAD authentication etc works smoothly) but leads to another problem. As soon as application loads in MS Teams after some seconds, MS Teams environment displays a message "There was a problem reaching this app". After spending some hours we found the exact reason i.e. MS Teams environment expects some confirmation from the loaded application but as our Open WebUI application don't have Teams App SDK required scripts so it was showing the error message (although despite this error message our application used to work smoothly).

For some reasons we're not supposed to customise our Open WebUI application, so as a workaround we decided to update the html response on the fly only for those requests which are generated from MS Teams i.e. if request headers User Agent contains "Teams" then intercept the response and add the Teams SDK specific scripts in HTML body so that it may notify the MS Teams Environment.

We used envoy script and deployed it on our pod which check the user agent and adds following script conditionally:

 <script src="https://res.cdn.office.net/teams-js/2.22.0/js/MicrosoftTeams.min.js"
      integrity="sha384-Wasdsadsad/sadasdae3e/ewqtr+werwerw/Qdx"
      crossorigin="anonymous"></script>

<script type="text/javascript">
      microsoftTeams.initialize(() => {
        console.log("Teams SDK initialized.");
        microsoftTeams.appInitialization.notifyAppLoaded();
        console.log("App loaded.");
        microsoftTeams.appInitialization.notifySuccess();
        console.log("Success notified.");
      });
    </script>

This is the same script that is added by Teams Toolkit as soon as we create a static tabs application. 

Here is another related community post that we created in our journey to fix the issues: 
https://learn.microsoft.com/en-us/answers/questions/2139367/there-was-a-problem-reaching-this-app-static-tabs?comment=question#comment-1876584

Hope, it'll help someone. 

Thanks

**Reply by dawnmarch** [reply]
@AhmadAhsan Thanks for you help. After adding the script, it did solve the "There was a problem reaching this app" prompt issue, and there are no problems in the Windows Teams client. That's great!
However, I noticed that the MAC Teams client still displays the following error. 

> Refused to display 'https://login.microsoftonline.com/' in a frame because it set 'X-Frame-Options' to 'deny'.

I saw that you mentioned earlier we don't need an iframe, but in the Teams client, the staticTab URL should be embedded via an iframe. I'm not sure why there are differences between client versions on different operating systems. How can this be resolved? Could you help me? Thank you.

**Reply by dawnmarch** [reply]
My contentUrl is configured to the address https://chat.xxx.com/oauth/microsoft/login, which redirects to https://login.microsoftonline.com/ to complete the OAuth request. I understand that this flow is necessary, but it seems that https://login.microsoftonline.com detects the current request and returns X-Frame-Options set to deny, preventing the process from continuing.

I'm a bit confused as to why everything works normally on the Windows client, but there's an issue on the MAC client.

**Reply by AhmadAhsan** [reply]
hi @dawnmarch 
If using web browser, you access your Open WebUI application using https://chat.xxx.com/ and SSO is working fine then as per my understanding your contentUrl should be set to the same [https://chat.xxx.com/]. 
Once your application is loaded in Teams Environment it should work without any issue, you don't need to mention the SSO login address. Double check if your [validDomains](https://learn.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema#validdomains) value is correct.

As your web application is not using any iframe so you shouldn't have this error:
`Refused to display 'https://login.microsoftonline.com/' in a frame because it set 'X-Frame-Options' to 'deny'.`

but in one case, if you previously tried with iframe and now Teams is showing you cached version.

**Reply by dawnmarch** [reply]
@AhmadAhsan Thanks for your quick reply. Regarding contentUrl, I also tried the homepage address.
My current phenomenon is: 1) In Teams' Windows APP, everything works fine 2) In Teams' Mac APP, it cannot be displayed because of the reasons I mentioned above.
I haven't tried the iframe method in the past, and I have always used the static tab setting, so the cache reason does not seem to be the reason.

1) I checked the http, the value of sec-fetch-dest header is iframe
2) Comparing the Mac APP and Windows APP requests, I found that the deny is returned in the Microsoft login http request of the Mac APP, and the Windows one is normal, so I am not sure whether Microsoft has restricted the Mac APP request.

So I would like to ask if both Mac and Windows Teams App are work fine?

**Reply by AhmadAhsan** [reply]
@dawnmarch in that case seems like issue is due to Teams Mac App, I suggest you to post the issue in relevant community:
https://learn.microsoft.com/en-us/answers/tags/339/office-teams-development

Try to be descriptive, in case you get some fixes, I'll appreciate if you can update this thread as well. Thanks

**Reply by zyborc** [reply]
Hi @AhmadAhsan 
 thanks for the information etc.
 
Could you eloborate a little more where you implmented exactly the script tag? Currently im running on the new MS Teams Client into the problem that the iframe directly goes to login.microsoftonline.com and there is there i got the error:Refused to display 'https://login.microsoftonline.com/' in a frame because it set 'X-Frame-Options' to 'deny'.

I configured the openwebui exactly like in the post with the "none" settings for cookies and merge user email etc. In a normal Iframe it works. But not with the new MS Teams client.

**Reply by arunbugkiller** [reply]
Hi @AhmadAhsan, the information is very helpful. I am able to resolve the "Problem Reaching App" issue, but facing issue with "Microsoft Login" functionality. SSO is working fine through browser but when trying to access through MSTeam, I can see the below error in OpenWebUI container logs 

2025-04-05 05:09:33.472 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 18.191.116.100:0 - "GET /oauth/microsoft/login HTTP/1.1" 302 - {}
2025-04-05 05:09:36.346 | WARNING  | open_webui.utils.oauth:handle_callback:237 - OAuth callback error: mismatching_state: CSRF Warning! State not equal in request and response. - {}
2025-04-05 05:09:36.346 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 18.191.116.100:0 - "GET /oauth/microsoft/callback?code=1.AS0AajkibDrW_k2lAUxCdjaHVVYliqvgoihNlRwZk50PG64tACstAA.AgABBAIAAABVrSpeuWamRam2jAF1XRQEAwDs_wUA9P_Hi8NxWjvoyb2qDblgMDs5M6RH94024LCoRIHd5NEdB80p44uVxRvFMLTo9ydnN7SFL8Wc4CssbBa3pJMMAWxVR4_QfGFts0GFMeLpKS7J0tyV2Bgs8ANLUQHEKt4YtRJMq7rrfKtaT6HO0zPHVQjvRokNPImYXlBnlI0AIXmR-cT1DMwuRV0-tQAmj9Ut3G7xw3wHHzZuMujwXYpBRfB27AcG6r83CuQ4x350hALYFuH8vjmAlWObSfEje9HTxOFV3XBg8WTLYEpJHSwTOxpbdDZIR3vxWN6Lz-o-tIPaVbC3GUp1bVXtoQm3QeM2iSRtdK_d7VnrHHMxOZ6Sz2UF4t26wZlyhWwgR_fI4zikiU7MyR5n2Y8xAMw0lwl_lNQgvvoGraGq9FA8UFm8Q87j7IHq4UctXiSWThpoVqdCB_qH_h295V-cF6r_dCyUxI3n4SquiVbzxQRUOtE07Bo_ivktLum170fnhuG9y_bD-xAOT_F8pRP0wuAE873AiZk_IHudmhQdyFHH9erpNzIT_sFlMRcRgdOVzGezTjyjAV1plVX1QbSVvK1_lQjIXmZql_xB-2p8U3kGlXh_F8xcWjzi4Enh-Tx5JwR34h_OWiP-_fPnxmEY9FPdy_EGWbbnK0opByz2jOGS-63dJrvIBrClgdLc0iJUn8skfxSSxgMG5BmLJ6fphCYSIaZgbu0AMtZORZ3ajB2cVr8VbYEjk48t92bi1b7knViYSCo90xfljXBEEp5HajGNZLsiFldosxboNB8BpktsWYLQy4DQwGxU8dAlxgLmW6u1qNghKVjzTKeS_W5eePc8p_SjUdSFuJgtELaiBPTqAd5TVPjNLmvFA5LXl7nmWg2EuSYe1JSYvXPiKBFn1R5V_7yP6lb_96GKQBbcdrHnOFnI0G0Cn4En-YhvrRqk-OXXqQmRaGBF9WP4yB5dRq9_0DXJCAjeUV3InWfD97jUVfk&state=GI5eyoy7hHgYXoIubYcaLy31taWwOu&session_state=00206129-e73f-f073-f99c-9fe102fec434 HTTP/1.1" 400 - {}

**Reply by leoo83** [reply]
Did you fix the error? I’m getting the same thing as you

**Comment by hanyEthan** [comment]
thank you @AhmadAhsan, according to this status, we can close this thread.

**Comment by 3lackdev** [comment]
Using the teams-js function, call popup windows Authenticator  like this:

1. create new function [openMicrosoftPopup]

- 

- [ ] const openMicrosoftPopup = async () => {

		try {
			await app.initialize();
			const resultToken = await authentication.authenticate({
				url: `${WEBUI_BASE_URL}/oauth/microsoft/login`,
				width: 600,
				height: 600
			});
			// console.log('Auth success, token:', resultToken);

			const sessionUser = await getSessionUser(resultToken);
			await setSessionUser(sessionUser);

			window.location.reload();
		} catch (error) {
			console.error('Auth failure:', error);
			window.location.href = `${WEBUI_BASE_URL}/oauth/microsoft/login`;
		}};

2. setup callback in [checkOauthCallback] when success will closed windows and return reload main page

- [ ] const checkOauthCallback = async () => {

 your code......

		try {
			await app.initialize();
			authentication.notifySuccess(token);
			return;
		} catch {
			authentication.notifyFailure('NotifyFailed');
			return;
		}
	};

3. Set button to on:click={openMicrosoftPopup}

**Comment by davejohnb** [comment]
Hi everybody, heeeelp! :) Since days we are trying to fix the problem of not being able to login with Microsoft through MS Teams (where as in the browser it works well). We are also getting the same message: 
`Refused to display 'https://login.microsoftonline.com/' in a frame because it set 'X-Frame-Options' to 'deny'.`

In the manifest.json we have the following settings and allowed domains set, but didn't help.
<img width="467" alt="Screenshot 2025-06-05 at 12 00 24 AM" src="https://github.com/user-attachments/assets/4c02447d-851d-4619-8845-e0e2908ef9b9" />

@AhmadAhsan @hanyEthan @3lackdev @dawnmarch maybe one of you can help us?

**Reply by asiedler** [reply]
Hi maybe this information can help you:

i was able to add openwebui via ms teams app and login works in every application (browser / Client / ios)
For that i got the added the domains as validDomains;
"login.microsoftonline.com",
"xxx.sharepoint.com"

i also created a new Azure Client App registration and added the WebApplicationInfo in the manifest

"webApplicationInfo": {
    "id": "ClientID",
    "resource": "api://OPENWEBUIURL/ClientID"
  }
In the azure app reg i have just User.Read permission and exposed an API and defined the application ID URI like in the example above. There i added a new scope like access_as_user

We host our Openwebui in a docker container inside of an Azure App Service. There we enable authentication etc.
In the Identity provider settings i added the created scope of the ms teams app registration to the allowed token audiences and enable the client requrements so that tis app can talk to the app services.

In the openwebui added the CONTENT_SECURITY_POLICY and added the urls from microsoft:
frame-ancestors 'self' *.microsoft.com *.sharepoint.com *.onedrive.com *.teams.microsoft.com OPENWEBUIURL *.cloud.microsoft

I did a lot of try and error but i think these settings will get it done

**Reply by bilalnazirraja** [reply]
Hi, I followed your steps, but i encountered a 500 error when accessing the OWUI MS Teams app. However, When i access the container app directly and refresh it a few times, i sometimes get a 500 error, and other times it shows the Microsoft login pop-up for OWUI. However, after signing in, i receive an error from Microsoft instead of being redirected properly. After logging in, though, it just displayed a white screen.
 Here's what I did based on your instructions:
•	I followed your steps and added the validDomains you listed.
•	I also created a new Azure client app.
•	In my Teams app's manifest.json, I added the webApplicationInfo section as follows:

"webApplicationInfo": {
  "id": "ClientID",
  "resource": "api://OPENWEBUIURL/ClientID"
}

•	I then created an Azure app registration, granted the User.Read permission, and exposed the API.

![Screenshot (23)](https://github.com/user-attachments/assets/303c5a2f-14ee-49cc-96a7-4c36b1efeea4)


I am running OWUI in a Docker container hosted inside an Azure App Service. I have enabled authentication and added the scope from the MS Teams app registration to the identity provider settings. However, I am still unsure if our configuration is correct for the authentication part.
Specifically, we added the created scope to the allowed token audiences and enabled the client requirements, similar to what you described. Could you provide a bit more detail on your exact configuration steps in this section?
We also added the same environment variables as yours.


![Screenshot (24)](https://github.com/user-attachments/assets/5cc2192e-b169-44f4-b291-32f7e1244783)

**Reply by asiedler** [reply]
Hi,
at first make sure you have this settings in the azure app service:
![image](https://github.com/user-attachments/assets/be2a1897-70e1-49de-a1f4-9da7694a5bbe)
If you dont allow unauthenticated request, then it wont work in teams. The MS Teams App display the openwebui in an **IFRAME** and it tries to redirect the user to microsoftonline login which will deny it because of iframe use. For that i didnt find any solution yet with my knowlede.

More or less you need to follow the sso guide for MS Teams Apps. you need to define the correct Redirect URL for the Web and for the SPA:
for example add https://localhost to the SPA redirect URL as well as a new login page that you need to create for the ms teams client login. You cant use the "normal" login from openwebui because popups dont work in ms teams client. They just work in the browser! You need to change the current openwebui code and add the teamsjs library: https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/using-teams-client-library?tabs=javascript%2Cmanifest-toolkit#authentication
There are a lot of examples how to do it online. Here are a small glance at what i did in the code in our openwebui repo:

For examlpe i added this code in the src/routes/+layout.svelte
`if(!token) {
					try {
						await microsoftTeams.app.initialize().then( async () => {
						await microsoftTeams.authentication.authenticate({
								url: "https://openwebui/oauth/microsoft/login", // Your backend authentication URL
								width: 600,
								height: 535,
							})
							.then((result) => {
								console.log("Authentication successful!", result);
								token = localStorage.token
								// Handle the result, such as storing the token or updating the UI
							})
							.catch((error) => {
								console.error("Authentication failed", error);
								// Handle the error, such as displaying an error message to the user
							});	
						}).catch(() => {
							console.log("no teams")
						});
					} catch {
						console.log("no ms teams client enviroment")
						token = null
					}
					
				}`

In the src/routes/auth/+page.svelte i added this to close the new teams popup:
`await microsoftTeams.app
				.initialize()
				.then(async () => {
					try {
						microsoftTeams.authentication.notifySuccess("Success - close now please");
						/*
						await microsoftTeams.app.initialize().then( async () => {
							microsoftTeams.authentication.notifySuccess(token);
						}).catch(() => {
							console.log("no teams")
						});*/
					} catch {
						console.log('got already token from popup and its already closed');
					}
				})
				.catch(() => {
					console.log('no teams');
				});`

**TO continure the setup for the app Reg:**
For the Teams App we use the same App reg we use for the Azure App service for identity provider.
You need to enable  Acces and ID Token in the Authentication settings as well.
Please make sure that you have all correct redirect url:
https://OPENWEBUI/oauth/microsoft/callback
https://OPENWEBUI/.auth/web/redirect
http://localhost:8080/oauth/microsoft/callback
https://OPENWEBUI/auth
https://OPENWEBUI/.auth/login/aad/callback

For API permission we use these settings as well:
![image](https://github.com/user-attachments/assets/66fedade-03f6-4405-bd6d-d9e6c6b937d3)
make sure you have the correct scope setup in your Azure App Service Enviroment as well
We use OneDrive etc as well.

What i saw on your Screenshot is that i think you are missing the Client Applications that can use your scope. Here you can find a list and guide where to add them:
https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/bot-sso-register-aad?tabs=windows#to-configure-an-authorized-client-application

![image](https://github.com/user-attachments/assets/3d62739e-dbf1-4dab-9f26-0280eac2ac57)


Be sure that you have setup all openwebui important configs. I Think these settings might help:
`
[
  {
    "name": "CONTENT_SECURITY_POLICY",
    "value": "frame-ancestors 'self' *.microsoft.com *.sharepoint.com *.onedrive.com *.teams.microsoft.com OPENWEBUIURL  *.cloud.microsoft",
    "slotSetting": false
  },
  {
    "name": "ENABLE_LOGIN_FORM",
    "value": "False",
    "slotSetting": false
  },
  {
    "name": "ENABLE_OAUTH_SIGNUP",
    "value": "True",
    "slotSetting": false
  },
  {
    "name": "ENABLE_SIGNUP",
    "value": "False",
    "slotSetting": false
  },
  {
    "name": "MICROSOFT_OAUTH_SCOPE",
    "value": "openid email profile Files.Read User.Read Files.ReadWrite offline_access",
    "slotSetting": false
  },
  {
    "name": "OAUTH_MERGE_ACCOUNTS_BY_EMAIL",
    "value": "True",
    "slotSetting": false
  },
  {
    "name": "OAUTH_PROVIDERS",
    "value": "microsoft",
    "slotSetting": false
  },
  {
    "name": "OPENID_PROVIDER_URL",
    "value": "https://login.microsoftonline.com/{tennant_id}/.well-known/openid-configuration",
    "slotSetting": false
  },
  {
    "name": "WEBUI_AUTH_COOKIE_SAME_SITE",
    "value": "none",
    "slotSetting": false
  },
  {
    "name": "WEBUI_AUTH_COOKIE_SECURE",
    "value": "True",
    "slotSetting": false
  },
  {
    "name": "WEBUI_SESSION_COOKIE_SAME_SITE",
    "value": "none",
    "slotSetting": false
  },
  {
    "name": "WEBUI_SESSION_COOKIE_SECURE",
    "value": "True",
    "slotSetting": false
  }
]`
